src/smartactions/SchedulingBase.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import "modulekit/core/sessionKey/ISessionValidationModule.sol";
+import { IERC20 } from "forge-std/interfaces/IERC20.sol";
+import { IERC7579Execution } from "modulekit/Accounts.sol";
+import { ERC7579ExecutorBase } from "modulekit/Modules.sol";
+
+abstract contract SchedulingBase is ERC7579ExecutorBase {
+    error InvalidExecution();
+
+    event ExecutionAdded(address indexed smartAccount, bytes32 indexed executionHash);
+
+    event ExecutionTriggered(address indexed smartAccount, bytes32 indexed executionHash);
+
+    event ExecutionCancelled(address indexed smartAccount, bytes32 indexed executionHash);
+
+    mapping(address smartAccount => mapping(bytes32 jobHash => ExecutionConfig)) internal
+        executionLog;
+
+    struct ExecutionConfig {
+        uint48 executeInterval;
+        uint16 numberOfExecutions;
+        uint16 numberOfExecutionsCompleted;
+        uint48 startDate;
+        bool isCancelled;
+        uint48 lastExecutionTime;
+        bytes executionData;
+    }
+
+    modifier canExecute(bytes32 jobHash) {
+        ExecutionConfig storage executionConfig = executionLog[msg.sender][jobHash];
+
+        if (executionConfig.isCancelled) {
+            revert InvalidExecution();
+        }
+
+        if (executionConfig.lastExecutionTime + executionConfig.executeInterval < block.timestamp) {
+            revert InvalidExecution();
+        }
+        if (executionConfig.numberOfExecutionsCompleted >= executionConfig.numberOfExecutions) {
+            revert InvalidExecution();
+        }
+        if (executionConfig.startDate > block.timestamp) {
+            revert InvalidExecution();
+        }
+
+        _;
+    }
+
+    // abstract methohd to be implemented by the inheriting contract
+    function executeOrder(bytes32 executionHash) external virtual;
+
+    function createExecution(bytes memory data) internal {
+        (
+            uint48 executeInterval,
+            uint16 numberOfExecutionsCompleted,
+            uint16 numberOfExecutions,
+            uint48 startDate,
+            ,
+            uint48 lastExecutionTime,
+            bytes memory executionData
+        ) = abi.decode(data, (uint48, uint16, uint16, uint48, bool, uint48, bytes));
+
+        bytes32 executionHash = keccak256(executionData);
```

Comment:
not sure if I like this:

lets say we want to have scheduled transactions of arbitrary contract calls. maybe a mint.


so the scheduled transaction lead to:

executionData = $nftContract + $mintFunction

but since the hash is calculated from the executionData only, scheduling two mints (at different execution times) would actually result in in the same executionHash.

the function doesnt check, if the executionHash already exists, so the 2nd schedule would just overwrite the first. and also nuking the `numberOfExecution` log if any were done on the 1st schedule already